### ISSUE TITLE
php -r error in cpanel_uapi.sh - Error in argument 1, char 2: option not found r

### ISSUE THREAD

**Issue by williatf** [Neilpang/acme.sh#1328]
Steps to reproduce
------------------
Running cron job with GoDaddy cPanel:
`~/.acme.sh/acme.sh --deploy --domain <domain> --deploy-hook cpanel_uapi`

I get an error stating that the system could not parse the certificate.

Traced the error to cpanel_uapi.sh lines 38 & 39:
```
  _cert=$(php -r "echo urlencode(\"$_certstr\");")
  _key=$(php -r "echo urlencode(\"$_keystr\");")
```

The php version that cron is using is CGI-FCGI, which does not have the -r option:

$  which -a php
```
/usr/local/bin/php
/usr/bin/php
```

$  /usr/bin/php -v
`PHP 5.6.32 (cgi-fcgi) (built: Dec  4 2017 12:21:23)`

$  /usr/bin/php -h
```
Usage: php [-q] [-h] [-s] [-v] [-i] [-f <file>]
       php <file> [args...]
  -a               Run interactively
  -b <address:port>|<port> Bind Path for external FASTCGI Server mode
  -C               Do not chdir to the script's directory
  -c <path>|<file> Look for php.ini file in this directory
  -n               No php.ini file will be used
  -d foo[=bar]     Define INI entry foo with value 'bar'
  -e               Generate extended information for debugger/profiler
  -f <file>        Parse <file>.  Implies `-q'
  -h               This help
  -i               PHP information
  -l               Syntax check only (lint)
  -m               Show compiled in modules
  -q               Quiet-mode.  Suppress HTTP Header output.
  -s               Display colour syntax highlighted source.
  -v               Version number
  -w               Display source with stripped comments and whitespace.
  -z <file>        Load Zend extension <file>.
  -T <count>       Measure execution time of script repeated <count> times.
```

I fixed it by hard-coding "/usr/local/bin/php" as the php version to use.
**Extract of cpanel_uapi.sh**
```
  if ! _exists /usr/local/bin/php; then
    _err "The command php is not found."
    return 1
  fi
  # read cert and key files and urlencode both
  _certstr=$(cat "$_ccert")
  _keystr=$(cat "$_ckey")
  _cert=$(/usr/local/bin/php -r "echo urlencode(\"$_certstr\");")
  _key=$(/usr/local/bin/php -r "echo urlencode(\"$_keystr\");")
```

**Possible solution is to trap the error (check for -r option) and/or make the php version an optional config param.**

Debug log
-----------------
```
[Mon Mar  5 13:35:01 MST 2018] LE_WORKING_DIR='<path>/.acme.sh'
[Mon Mar  5 13:35:01 MST 2018] Using config home:<path>/.acme.sh
[Mon Mar  5 13:35:01 MST 2018] ACME_DIRECTORY='https://acme-v01.api.letsencrypt.org/directory'
[Mon Mar  5 13:35:01 MST 2018] _ACME_SERVER_HOST='acme-v01.api.letsencrypt.org'
[Mon Mar  5 13:35:01 MST 2018] DOMAIN_PATH='<path>/.acme.sh/<domain>'
[Mon Mar  5 13:35:01 MST 2018] _deployApi='<path>/.acme.sh/deploy/cpanel_uapi.sh'
[Mon Mar  5 13:35:01 MST 2018] _cdomain='<domain>'
[Mon Mar  5 13:35:01 MST 2018] _ckey='<path>/.acme.sh/<domain>/<domain>.key'
[Mon Mar  5 13:35:01 MST 2018] _ccert='<path>/.acme.sh/<domain>/<domain>.cer'
[Mon Mar  5 13:35:01 MST 2018] _cca='<path>/.acme.sh/<domain>/ca.cer'
[Mon Mar  5 13:35:01 MST 2018] _cfullchain='<path>/.acme.sh/<domain>/fullchain.cer'
[Mon Mar  5 13:35:02 MST 2018] _cert='Usage: php [-q] [-h] [-s] [-v] [-i] [-f <file>]
       php <file> [args...]
  -a               Run interactively
  -b <address:port>|<port> Bind Path for external FASTCGI Server mode
  -C               Do not chdir to the script's directory
  -c <path>|<file> Look for php.ini file in this directory
  -n               No php.ini file will be used
  -d foo[=bar]     Define INI entry foo with value 'bar'
  -e               Generate extended information for debugger/profiler
  -f <file>        Parse <file>.  Implies `-q'
  -h               This help
  -i               PHP information
  -l               Syntax check only (lint)
  -m               Show compiled in modules
  -q               Quiet-mode.  Suppress HTTP Header output.
  -s               Display colour syntax highlighted source.
  -v               Version number
  -w               Display source with stripped comments and whitespace.
  -z <file>        Load Zend extension <file>.
  -T <count>       Measure execution time of script repeated <count> times.'
[Mon Mar  5 13:35:02 MST 2018] _key='Usage: php [-q] [-h] [-s] [-v] [-i] [-f <file>]
       php <file> [args...]
  -a               Run interactively
  -b <address:port>|<port> Bind Path for external FASTCGI Server mode
  -C               Do not chdir to the script's directory
  -c <path>|<file> Look for php.ini file in this directory
  -n               No php.ini file will be used
  -d foo[=bar]     Define INI entry foo with value 'bar'
  -e               Generate extended information for debugger/profiler
  -f <file>        Parse <file>.  Implies `-q'
  -h               This help
  -i               PHP information
  -l               Syntax check only (lint)
  -m               Show compiled in modules
  -q               Quiet-mode.  Suppress HTTP Header output.
  -s               Display colour syntax highlighted source.
  -v               Version number
  -w               Display source with stripped comments and whitespace.
  -z <file>        Load Zend extension <file>.
  -T <count>       Measure execution time of script repeated <count> times.'
[Mon Mar  5 13:35:02 MST 2018] Error in deploying certificate:
[Mon Mar  5 13:35:02 MST 2018] --- 
apiversion: 3
func: install_ssl
module: SSL
result: 
  data: ~
  errors: 
    - "The system could not parse the certificate because of an error: The certificate text was not valid."
  messages: ~
  metadata: {}

  status: 0
[Mon Mar  5 13:35:02 MST 2018] Error deploy for domain:<domain>
[Mon Mar  5 13:35:02 MST 2018] Deploy error.

**Comment by Neilpang** [comment]
can you please send a PR ?

**Comment by williatf** [comment]
What’s a PR?

Sent from my iPhone

On Mar 6, 2018, at 6:37 AM, neil <notifications@github.com> wrote:

can you please send a PR ?

**Comment by williatf** [comment]
Ah. Pull Request. Never mind. Let me work out the new code and I’ll send a PR. 

Todd


On Mar 6, 2018, at 7:25 AM, Todd Williams <williatf@gmail.com> wrote:

What’s a PR?

Sent from my iPhone

On Mar 6, 2018, at 6:37 AM, neil <notifications@github.com> wrote:

can you please send a PR ?

**Comment by NTMMFTS** [comment]
Same issue for me from godaddy cron job, but the same --deploy command works from the ssh shell and logs the successful deployment.

~/.acme.sh/acme.sh --deploy -d [my actual domain] --deploy-hook cpanel_uapi --log ~/.acme.sh/logs/deploy.log

**Comment by Abukamel** [comment]
I think the problem is that cPanel wants the certificate text to contain `\n` characters in it, it can't parse it without new lines, I knew that from trying to manually add the certificate via GUI away from uapi cli.

**Comment by Abukamel** [comment]
solution was to change the uapi call command inside the script to use the certificate string as it is without manipulating it via url_encode function
`_response=$(uapi --user="$DEPLOY_CPANEL_USER" SSL install_ssl domain="$_cdomain" cert="$_certstr" key="$_keystr")`

**Comment by santerikannisto** [comment]
I made a universal bug fix #1648 but currently it is NOT POSIX so it can not be merged. The fix works at least for me. I am having hard time finding a POSIX compliant way for string indexing:

`_print_offset="${1:_offset:1}"`

I need to get a substring of 1 character to urlencode it. The other issue, the for loop, I tackled by changing it to compliant while loop. Any ideas would be greatly appreciated.

Anyway, you can use the new code in my pull request as a temporary workaround.

**Comment by santerikannisto** [comment]
Done. Here is the fix: #1650

I rewrote the urlencode function using sed so now it is fully POSIX and not more php dependency. The script worked on the servers I am maintaining but that does not guarantee anything. Please let me know if you run into any issues  I haven't used sed for 18 years.

Happy hacking,

Santeri

**Comment by santerikannisto** [comment]
Here is my last PR on GitHub: #1664 . I found a POSIX compatibility issue from sed extension (N) I had used in my earlier PR. This one should be fully portable.

If you have any issues or questions regarding my code, feel free to contact me at my [discussion forum](https://forum.webseodesigners.com/web-design-seo-and-hosting-f16/). I prefer giving public answers as often I get the same questions asked time and again.

My projects are in [GitLab](https://gitlab.com/santerikannisto) including the portable urlencode function I wrote for this bug fix to replace php. I hope everything will go fine with GitHub for you and wish you all the best. Thanks for the great script @Neilpang . Take care.
